flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,331 @@
+
+        private final List<ParserRuleContext> extractedSingleBlocks = new ArrayList<>();
+
+        private final String context;
+
+        private final CommonTokenStream tokenStream;
+
+        private final TokenStreamRewriter rewriter;
+
+        private final String parameters;
+
+        private int counter = 0;
+
+        private int bscCounter = 0;
+
+        private BlockStatementVisitor(String code, String context, String parameters) {
+            this.tokenStream = new CommonTokenStream(new JavaLexer(CharStreams.fromString(code)));
+            this.rewriter = new TokenStreamRewriter(tokenStream);
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        private BlockStatementVisitor(
+                TokenStreamRewriter rewriter, String context, String parameters) {
+            this.tokenStream = null;
+            this.rewriter = rewriter;
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        @Override
+        public Void visitStatement(StatementContext ctx) {
+
+            if (ctx.getChildCount() == 0) {
+                return null;
+            }
+
+            if (ctx.WHILE() != null) {
+                for (StatementContext statement : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_whileBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(statement);
+                }
+            } else if (ctx.IF() != null) {
+                for (StatementContext sc : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_ifBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(sc);
+                }
+
+            } else if (ctx.ELSE() != null) {
+                System.out.println(CodeSplitUtil.getContextString(ctx));
+            } else {
+                if (ctx.block() != null) {
+                    for (BlockStatementContext bsc : ctx.block().blockStatement()) {
+                        if (bsc.statement() != null
+                                && (bsc.statement().IF() != null
+                                        || bsc.statement().ELSE() != null
+                                        || bsc.statement().WHILE() != null)) {
+
+                            String localContext =
+                                    String.format("%s_%d", this.context, bscCounter++);
+                            BlockStatementVisitor visitor =
+                                    new BlockStatementVisitor(rewriter, localContext, parameters);
+                            visitor.addBlocks(extractedSingleBlocks);
```

Comment:
Why are you adding blocks to the recursively created `BlockStatementVisitor`? This will cause the problem that blocks in `extractedSingleBlocks` might not be the same block.

Reply:
I'm assuming you are wondering about case similar to the example below: 
```
public void myFun1(int[] a, int[] b) throws RuntimeException {
        if (a[0] == 0) {
            a[11] = b[0];
            a[12] = b[0];
            if (a[2] == 0) {
                a[21] = 1;
                a[22] = 1;
            } else {
                a[23] = b[2];
                a[24] = b[2];
            }

            a[13] = b[0];
            a[14] = b[0];
        } else {
            a[0] = b[0];
            a[1] = b[1];
            a[2] = b[2];
        }
    }
```

In this case, the rewritten code will look like this:
```
public void myFun1(int[] a, int[] b) throws RuntimeException {
  
	if (a[0] == 0) {
		myFun1_rewriteGroup1(a, b);
  
		myFun1_ifBody0(a, b);
	} else {
		myFun1_ifBody1(a, b);
	}
}
  
  void myFun1_ifBody0_0_ifBody0(int[] a, int[] b) throws RuntimeException {
	a[21] = 1;
	a[22] = 1;
  }
  
  void myFun1_ifBody0_0_ifBody1(int[] a, int[] b) throws RuntimeException {
	a[23] = b[2];
	a[24] = b[2];
  }
  
  void myFun1_rewriteGroup1(int[] a, int[] b) throws RuntimeException {
	myFun1_ifBody0_0(a, b);
	if (a[2] == 0) {
		myFun1_ifBody0_0_ifBody0(a, b);
	} else {
	myFun1_ifBody0_0_ifBody1(a, b);
	}
  }
  
void myFun1_ifBody0_0(int[] a, int[] b) throws RuntimeException {
	a[11] = b[0];
	a[12] = b[0];
  }
  
  void myFun1_ifBody1(int[] a, int[] b) throws RuntimeException {
	a[0] = b[0];
	a[1] = b[1];
	a[2] = b[2];
  }
  
  void myFun1_ifBody0(int[] a, int[] b) throws RuntimeException {
	a[13] = b[0];
	a[14] = b[0];
  }
```

First time when we hit `for (BlockStatementContext bsc : ctx.block().blockStatement()) {` and its IF/ELSE/WHILE check will be for block:
```
if (a[2] == 0) {
	a[21] = 1;
	a[22] = 1;
} else {
	a[23] = b[2];
	a[24] = b[2];
}
```
For this block we will create a new Visitor and pass `extractedSingleBlocks` array that contains single line statements 
`a[11] = b[0];` and `a[12] = b[0];`.

When we call `visitStatement` on this new visitor, the ` } else if (ctx.IF() != null) {` will be true. The we will create new Visitors, one for TRUE and FALSE branch of that statement. Both with have empty `extractedSingleBlocks`.

So lets take a look at `myFun1_ifBody0` context.
At first we will have `a[11] = b[0];` and `a[12] = b[0];` added to the `extractedSingleBlocks` for this context.
Then we will go into ` `for (BlockStatementContext bsc : ctx.block().blockStatement()) {`` where we will create new visitor, with new context and we will pass `a[11] = b[0];` and `a[12] = b[0];` to there. Finally those will land under `myFun1_ifBody0_0` which happens to be this new created Visitor.

Then going back to `myFun1_ifBody0`, now it has its `extractedSingleBlocks` empty, since we just clear it. The next and finnal call for this context will be adding `a[13] = b[0];` and `a[14] = b[0];` to `extractedSingleBlocks` which are last statement to process for this block. 


For more details please take a look at my response for -> https://github.com/apache/flink/pull/21393#discussion_r1071233842